### ISSUE TITLE
Routing between VLANs does not work: no ARP request is sent

### ISSUE THREAD

**Issue by charlag** [Azure/sonic-buildimage#9470]
#### Description

Routing from the L3 port to the device in VLAN does not produce ARP packets in said VLAN. ARP table does not contain information about the device in VLAN until said device sends a packet towards the router.

#### Steps to reproduce the issue:
1. Apply config_db.json from below
2. Set up device A: on Ethernet68 with IP 10.0.0.2/24 and add route `ip route add 10.1.0.0/24 via 10.0.0.1`
3. Set up device B: on Ethernet0 with IP 10.1.0.2/24 and add route `ip route add 10.0.0.0/24 via 10.1.0.1`
4. Verify that neither of the devices is in the arp table: `show arp`. If they are there, remove them with `sudo ip -s -s neigh flush all `
5. Ping from device B to device A

#### Describe the results you received:
No response. Packets are reaching the router and are detectable with tcpdump but no ARP packets are sent via Ethernet0.

#### Describe the results you expected:
ARP packets are sent when exact match is not found for the packet destination and the hardware address of device B is learned.

#### Output of `show version`:

```
SONiC Software Version: SONiC.202106.56864-3f3dceb96
Distribution: Debian 10.11
Kernel: 4.19.0-12-2-amd64
Build commit: 3f3dceb96
Build date: Sun Dec  5 14:31:50 UTC 2021
Built by: AzDevOps@sonic-build-workers-000YIS

Platform: x86_64-mlnx_msn2010-r0
HwSKU: ACS-MSN2010
ASIC: mellanox
ASIC Count: 1
Serial Number: MT2119J17994
Model Number: MSN2010-CB2FO
Hardware Revision: A3
Uptime: 10:30:59 up 8 min,  2 users,  load average: 0.40, 0.65, 0.44

Docker images:
REPOSITORY                    TAG                      IMAGE ID            SIZE
docker-orchagent              202106.56864-3f3dceb96   ce6546197fd8        451MB
docker-orchagent              latest                   ce6546197fd8        451MB
docker-sflow                  202106.56864-3f3dceb96   ea61353eac42        433MB
docker-sflow                  latest                   ea61353eac42        433MB
docker-teamd                  202106.56864-3f3dceb96   0908f79ac7e4        433MB
docker-teamd                  latest                   0908f79ac7e4        433MB
docker-platform-monitor       202106.56864-3f3dceb96   0052e1f87551        749MB
docker-platform-monitor       latest                   0052e1f87551        749MB
docker-nat                    202106.56864-3f3dceb96   ce051ad0f5ea        435MB
docker-nat                    latest                   ce051ad0f5ea        435MB
docker-macsec                 202106.56864-3f3dceb96   acdb5e18ea81        436MB
docker-macsec                 latest                   acdb5e18ea81        436MB
docker-fpm-frr                202106.56864-3f3dceb96   abf2d124c869        451MB
docker-fpm-frr                latest                   abf2d124c869        451MB
docker-dhcp-relay             latest                   bd1a42ecade1        428MB
docker-sonic-mgmt-framework   202106.56864-3f3dceb96   f6421f6cf688        576MB
docker-sonic-mgmt-framework   latest                   f6421f6cf688        576MB
docker-snmp                   202106.56864-3f3dceb96   22bac2d54a69        463MB
docker-snmp                   latest                   22bac2d54a69        463MB
docker-lldp                   202106.56864-3f3dceb96   2d5d0ef2eac6        461MB
docker-lldp                   latest                   2d5d0ef2eac6        461MB
docker-sonic-telemetry        202106.56864-3f3dceb96   48a7d2a6fd5a        510MB
docker-sonic-telemetry        latest                   48a7d2a6fd5a        510MB
docker-database               202106.56864-3f3dceb96   4296b45487c6        421MB
docker-database               latest                   4296b45487c6        421MB
docker-router-advertiser      202106.56864-3f3dceb96   63b5dc3b90bc        421MB
docker-router-advertiser      latest                   63b5dc3b90bc        421MB
docker-syncd-mlnx             202106.56864-3f3dceb96   38f001bc78f6        1e+03MB
docker-syncd-mlnx             latest                   38f001bc78f6        1e+03MB
```

#### Additional information you deem important (e.g. issue happens only occasionally):

If the packets are sent from the control plane or are reaching router some other way then the ARP table is populated correctly. However when the routing is done via data plane then no ARP packets are produced.

config_db.json:

```json
{
    "BREAKOUT_CFG": {
        "Ethernet0": {
            "brkout_mode": "1x25G[10G]"
        },
        "Ethernet4": {
            "brkout_mode": "1x25G[10G]"
        },
        "Ethernet8": {
            "brkout_mode": "1x25G[10G]"
        },
        "Ethernet12": {
            "brkout_mode": "1x25G[10G]"
        },
        "Ethernet16": {
            "brkout_mode": "1x25G[10G]"
        },
        "Ethernet20": {
            "brkout_mode": "1x25G[10G]"
        },
        "Ethernet24": {
            "brkout_mode": "1x25G[10G]"
        },
        "Ethernet28": {
            "brkout_mode": "1x25G[10G]"
        },
        "Ethernet32": {
            "brkout_mode": "1x25G[10G]"
        },
        "Ethernet36": {
            "brkout_mode": "1x25G[10G]"
        },
        "Ethernet40": {
            "brkout_mode": "1x25G[10G]"
        },
        "Ethernet44": {
            "brkout_mode": "1x25G[10G]"
        },
        "Ethernet48": {
            "brkout_mode": "1x25G[10G]"
        },
        "Ethernet52": {
            "brkout_mode": "1x25G[10G]"
        },
        "Ethernet56": {
            "brkout_mode": "1x25G[10G]"
        },
        "Ethernet60": {
            "brkout_mode": "1x25G[10G]"
        },
        "Ethernet64": {
            "brkout_mode": "1x25G[10G]"
        },
        "Ethernet68": {
            "brkout_mode": "1x25G[10G]"
        },
        "Ethernet72": {
            "brkout_mode": "1x100G[50G,40G,25G,10G]"
        },
        "Ethernet76": {
            "brkout_mode": "1x100G[50G,40G,25G,10G]"
        },
        "Ethernet80": {
            "brkout_mode": "1x100G[50G,40G,25G,10G]"
        },
        "Ethernet84": {
            "brkout_mode": "1x100G[50G,40G,25G,10G]"
        }
    },
    "DEVICE_METADATA": {
        "localhost": {
            "hostname": "test",
            "hwsku": "ACS-MSN2010",
            "mac": "04:3f:72:5e:1b:80",
            "platform": "x86_64-mlnx_msn2010-r0",
            "docker_routing_config_mode": "split",
            "type": "LeafRouter"
        }
    },
    "INTERFACE": {
        "Ethernet68|10.0.0.1/24": {}
    },
    "LOOPBACK_INTERFACE": {
        "Loopback0|10.1.1.1/32": {}
    },
    "PORT": {
        "Ethernet0": {
            "admin_status": "up",
            "alias": "etp1",
            "index": "1",
            "lanes": "0",
            "mtu": "1500",
            "speed": "1000"
        },
        "Ethernet4": {
            "admin_status": "up",
            "alias": "etp2",
            "index": "2",
            "lanes": "4",
            "mtu": "1500",
            "speed": "25000"
        },
        "Ethernet8": {
            "admin_status": "up",
            "alias": "etp3",
            "index": "3",
            "lanes": "8",
            "mtu": "1500",
            "speed": "25000"
        },
        "Ethernet12": {
            "admin_status": "up",
            "alias": "etp4",
            "index": "4",
            "lanes": "12",
            "mtu": "1500",
            "speed": "25000"
        },
        "Ethernet16": {
            "admin_status": "up",
            "alias": "etp5",
            "index": "5",
            "lanes": "16",
            "mtu": "1500",
            "speed": "25000"
        },
        "Ethernet20": {
            "admin_status": "up",
            "alias": "etp6",
            "index": "6",
            "lanes": "20",
            "mtu": "1500",
            "speed": "25000"
        },
        "Ethernet24": {
            "admin_status": "up",
            "alias": "etp7",
            "index": "7",
            "lanes": "24",
            "mtu": "1500",
            "speed": "25000"
        },
        "Ethernet28": {
            "admin_status": "up",
            "alias": "etp8",
            "index": "8",
            "lanes": "28",
            "mtu": "1500",
            "speed": "25000"
        },
        "Ethernet32": {
            "admin_status": "up",
            "alias": "etp9",
            "index": "9",
            "lanes": "32",
            "mtu": "1500",
            "speed": "25000"
        },
        "Ethernet36": {
            "admin_status": "up",
            "alias": "etp10",
            "index": "10",
            "lanes": "36",
            "mtu": "1500",
            "speed": "25000"
        },
        "Ethernet40": {
            "admin_status": "up",
            "alias": "etp11",
            "index": "11",
            "lanes": "40",
            "mtu": "1500",
            "speed": "25000"
        },
        "Ethernet44": {
            "admin_status": "up",
            "alias": "etp12",
            "index": "12",
            "lanes": "44",
            "mtu": "1500",
            "speed": "25000"
        },
        "Ethernet48": {
            "admin_status": "up",
            "alias": "etp13",
            "index": "13",
            "lanes": "48",
            "mtu": "1500",
            "speed": "25000"
        },
        "Ethernet52": {
            "admin_status": "up",
            "alias": "etp14",
            "index": "14",
            "lanes": "52",
            "mtu": "1500",
            "speed": "1000"
        },
        "Ethernet56": {
            "admin_status": "up",
            "alias": "etp15",
            "index": "15",
            "lanes": "56",
            "mtu": "1500",
            "speed": "1000"
        },
        "Ethernet60": {
            "admin_status": "up",
            "alias": "etp16",
            "index": "16",
            "lanes": "60",
            "mtu": "1500",
            "speed": "1000"
        },
        "Ethernet64": {
            "admin_status": "up",
            "alias": "etp17",
            "index": "17",
            "lanes": "64",
            "mtu": "1500",
            "speed": "10000"
        },
        "Ethernet68": {
            "admin_status": "up",
            "alias": "etp18",
            "index": "18",
            "lanes": "68",
            "mtu": "1500",
            "speed": "1000"
        },
        "Ethernet72": {
            "admin_status": "up",
            "alias": "etp19",
            "index": "19",
            "lanes": "72,73,74,75",
            "mtu": "1500",
            "speed": "100000"
        },
        "Ethernet76": {
            "admin_status": "up",
            "alias": "etp20",
            "index": "20",
            "lanes": "76,77,78,79",
            "mtu": "1500",
            "speed": "100000"
        },
        "Ethernet80": {
            "admin_status": "up",
            "alias": "etp21",
            "index": "21",
            "lanes": "80,81,82,83",
            "mtu": "1500",
            "speed": "100000"
        },
        "Ethernet84": {
            "admin_status": "up",
            "alias": "etp22",
            "index": "22",
            "lanes": "84,85,86,87",
            "mtu": "1500",
            "speed": "100000"
        }
    },
    "VLAN": {
        "Vlan100": {
            "vlanid": "100"
        }
    },
    "VLAN_INTERFACE": {
        "Vlan100|10.1.0.1/24": {}
    },
    "VLAN_MEMBER": {
        "Vlan100|Ethernet0": {
            "tagging_mode": "untagged"
        }
    }
}
```

**Comment by dgsudharsan** [comment]
ip route commands don't program the route in hardware. In SONiC config commands should be used to configure static route. Please refer https://github.com/Azure/sonic-utilities/blob/master/doc/Command-Reference.md#static-routing-config-commands to configure static route

**Comment by dgsudharsan** [comment]
One more point, there is another issue with mellanox SDK for which fix is being worked on. Once the SDK fix is updated I will close this issue.

**Comment by charlag** [comment]
@dgsudharsan we do not to configure static routing (or any routing) on the SONIC device, this was just an example for connecting devices _to_ the router. Sonic device imports two subnets into the routing table from them just being directly connected.
The whole thing is a simplified example of our scenario where one side is inside VLAN and the other is coming via upstream and populated via BGP.

**Comment by dgsudharsan** [comment]
@charlag I see. Can you please add techsupport information to the issue? Please execute 'show techsupport' and attach the tar.gz file generated

**Comment by prsunny** [comment]
Not clear on this configuration:

```
Set up device A: on Ethernet68 with IP 10.0.0.2/24 and add route ip route add 10.1.0.0/24 via 10.0.0.1
Set up device B: on Ethernet0 with IP 10.1.0.2/24 and add route ip route add 10.0.0.0/24 via 10.1.0.1
```

You've a directly connected subnet 10.1.0.0/24 on Ethernet0. So why do you create a route to the same subnet but pointing to 10.0.0.1 via Ethernet68?

**Comment by dgsudharsan** [comment]
> Not clear on this configuration:
> 
> ```
> Set up device A: on Ethernet68 with IP 10.0.0.2/24 and add route ip route add 10.1.0.0/24 via 10.0.0.1
> Set up device B: on Ethernet0 with IP 10.1.0.2/24 and add route ip route add 10.0.0.0/24 via 10.1.0.1
> ```
> 
> You've a directly connected subnet 10.1.0.0/24 on Ethernet0. So why do you create a route to the same subnet but pointing to 10.0.0.1 via Ethernet68?

I had the same doubt but I believe the routes were created on the hosts rather than on SONiC.

**Comment by prsunny** [comment]
> I had the same doubt but I believe the routes were created on the hosts rather than on SONiC.

I see.. Since you mentioned static route on Sonic, I thought its on Sonic device.

Anyways, I don't find the config_db entry correct as well:

```
    "VLAN_INTERFACE": {
        "Vlan100|10.1.0.1/24": {}
    },

```
you must have one more entry as below:
```

    "VLAN_INTERFACE": {
         "Vlan100": {},
        "Vlan100|10.1.0.1/24": {}
    },
```

**Comment by prsunny** [comment]
Please use `'config interface'` command to add ip address and not directly edit the config_db.json

**Comment by charlag** [comment]
Hello again

We would like to apologize for taking your time, it is indeed a misconfiguration, we've found it when we reinstalled the system image to make a clean report.

Turns out our firewall was blocking forwarding of packets. We didn't discover this until now because when the ARP entry is there control plane wouldn't invoke the firewall and the packets would come through but it seems that when there's no ARP entry then the processing goes through control plane and blocks it. It took us a lot of time to discover this.

Apologies one more time.
